src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.ConnectionPools.cs

```diff
@@ -0,0 +1,347 @@
+                }
+
+                protected override void OnDisposed()
+                {
+                    _pools.Free(_serviceName, _connection);
+                }
+            }
+        }
+
+        private static class Connections
+        {
+            /// <summary>
+            /// call <paramref name="funcAsync"/> and retry up to <paramref name="timeout"/> if the call throws
+            /// <typeparamref name="TException"/>. any other exception from the call won't be handled here.
+            /// </summary>
+            public static async Task<TResult> RetryRemoteCallAsync<TException, TResult>(
+                Func<Task<TResult>> funcAsync,
+                TimeSpan timeout,
+                CancellationToken cancellationToken) where TException : Exception
+            {
+                const int retry_delayInMS = 50;
+
+                using (var pooledStopwatch = SharedPools.Default<Stopwatch>().GetPooledObject())
+                {
+                    var watch = pooledStopwatch.Object;
+                    watch.Start();
+
+                    while (watch.Elapsed < timeout)
+                    {
+                        cancellationToken.ThrowIfCancellationRequested();
+
+                        try
+                        {
+                            return await funcAsync().ConfigureAwait(false);
+                        }
+                        catch (TException)
+                        {
+                            // throw cancellation token if operation is cancelled
+                            cancellationToken.ThrowIfCancellationRequested();
+                        }
+
+                        // wait for retry_delayInMS before next try
+                        await Task.Delay(retry_delayInMS, cancellationToken).ConfigureAwait(false);
+
+                        ReportTimeout(watch);
+                    }
+                }
+
+                // operation timed out, more than we are willing to wait
+                ShowInfoBar();
+
+                // user didn't ask for cancellation, but we can't fullfill this request. so we
+                // create our own cancellation token and then throw it. this doesn't guarantee
+                // 100% that we won't crash, but this is at least safest way we know until user
+                // restart VS (with info bar)
+                using (var ownCancellationSource = new CancellationTokenSource())
+                {
+                    ownCancellationSource.Cancel();
+                    ownCancellationSource.Token.ThrowIfCancellationRequested();
+                }
+
+                throw ExceptionUtilities.Unreachable;
+            }
+
+            public static async Task<Stream> RequestServiceAsync(
+                HubClient client,
+                string serviceName,
+                HostGroup hostGroup,
+                TimeSpan timeout,
+                CancellationToken cancellationToken = default)
```

Comment:
should callers ever not pass this cancellation token?  i would not make default. too easy to screw up.  let people explicitly pass default if they watn.

Reply:
might later. but not now.

Reply:
OK.  Then make this non-default.